src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClients/FailOverSupportedConfigurationClient.cs

```diff
@@ -0,0 +1,256 @@
+            this._parallelRetryInterval = options.ParallelRetryInterval;
+        }
+
+        public Task<Response<ConfigurationSetting>> GetConfigurationSettingAsync(string key, string label = null, CancellationToken cancellationToken = default)
+        {
+            Task<Response<ConfigurationSetting>> func(ConfigurationClient client)
+            {
+                return client.GetConfigurationSettingAsync(key, label, cancellationToken);
+            }
+
+            return ExecuteWithFailOverPolicyAsync(func, cancellationToken);
+        }
+
+        public Task<Response<ConfigurationSetting>> GetConfigurationSettingAsync(ConfigurationSetting setting, bool onlyIfChanged = false, CancellationToken cancellationToken = default)
+        {
+            Task<Response<ConfigurationSetting>> func(ConfigurationClient client)
+            {
+                return client.GetConfigurationSettingAsync(setting, onlyIfChanged, cancellationToken);
+            }
+
+            return ExecuteWithFailOverPolicyAsync(func, cancellationToken);
+        }
+
+        public async Task<IEnumerable<ConfigurationSetting>> GetConfigurationSettingsAsync(SettingSelector selector, CancellationToken cancellationToken = default)
+        {
+            Page<ConfigurationSetting> settingsPage = null;
+            string continuationToken = null;
+            var result = new List<ConfigurationSetting>();
+
+            async Task<Page<ConfigurationSetting>> func(ConfigurationClient client)
+            {
+                IAsyncEnumerator<Page<ConfigurationSetting>> enumerator = client.GetConfigurationSettingsAsync(selector, cancellationToken)
+                                                                                .AsPages(continuationToken: continuationToken)
+                                                                                .GetAsyncEnumerator();
+
+                Page<ConfigurationSetting> page = enumerator != null && await enumerator.MoveNextAsync() ? enumerator.Current : null;
+                return page;
+            }
+
+            do
+            {
+                settingsPage = await ExecuteWithFailOverPolicyAsync(func, cancellationToken);
+                if (settingsPage != null)
+                {
+                    result.AddRange(settingsPage.Values);
+                    continuationToken = settingsPage.ContinuationToken;
+                }
+            }
+            while (settingsPage != null && settingsPage.ContinuationToken != null);
+
+            return result;
+        }
+
+        public void UpdateSyncToken(Uri endpoint, string syncToken)
+        {
+            this._clients.Single(clientAndState => clientAndState.Endpoint.Host.Equals(endpoint.Host)).Client.UpdateSyncToken(syncToken);
+        }
+
+        private async Task<T> ExecuteWithFailOverPolicyAsync<T>(Func<ConfigurationClient, Task<T>> funcToExecute, CancellationToken cancellationToken = default)
```

Comment:
Given the complexity of this code, I am leaning towards short-circuiting the case when there is only one client provided. There is a lot of overhead to go through (see how many tasks you have created) for a case there is nothing to fail over.

Reply:
I agree with you, but @jimmyca15 had a conflicting opinion when I had proposed this earlier.
I would love to hear his opinion about it too.

Reply:
my suggestion was to always use FailOverSupportedConfigurationClient. But as Zhenlan mentioned, it doesn't make sense to spin up a tasks used for parallelism in single client usage.